### DISCUSSION TITLE
Avoid having to export `tagName` for custom elements

### DISCUSSION THREAD

**Discussion by thepassle** [withastro/roadmap#175] [Proposal]
Currently when SSRing a custom element, its required to export a `tagName`, e.g.:

```js
class MyEl extends HTMLElement {}
export const tagName = 'my-el';
customElements.define(tagName, MyEl);
```

If you omit exporting the `tagName`, you'll get the following error:
![image](https://user-images.githubusercontent.com/17054057/164741567-adb7514b-b134-45c4-bb12-558f3bbd8de1.png)

if you dont export a `tagName` the `_props` looks like:
```js
{
  'client:visible': true,
  selected: '1',
  label: 'people',
  'client:component-hydration': 'visible',
  'client:component-path': null, // ❌
  'client:component-export': null, // ❌
  class: 'astro-YHQMMVYC'
}
```

if I do export a `tagName`, the `_props` look like:
```js
{
  'client:visible': true,
  selected: '1',
  label: 'people',
  'client:component-hydration': 'visible',
  'client:component-path': '/src/components/my-el.js', // ✅
  'client:component-export': 'tagName', // ✅
  class: 'astro-YHQMMVYC'
}
```

I tried searching the repo to see where `"client:component-path"` is coming from, but I didnt get any results, which is making me assume it's coming from the compiler. Maybe @natemoo-re can shed a light on this.

It'd be really nice to be able to either check the `customElements.define` call for the `tagName` if its a string (e.g.: `customElements.define('my-el' /* <- use this */, MyEl)`, or see if the class thats being passed to `.define` has a `static is` property (which holds the tagname) e.g.:
```js
class MyEl extends HTMLElement {
  static is = 'my-el';
}
```

or if both of those things are not the case, require a `tagName` to be exported.

It makes sense to use the `customElements.define` call as being an ‘export’, because it makes the component globally available. We also count `.define` calls as being exports in the [Custom Elements Manifest](https://github.com/webcomponents/custom-elements-manifest/blob/main/schema.d.ts#L137) schema community standard:
<img width="1326" alt="image" src="https://user-images.githubusercontent.com/17054057/164742721-4d9a9cab-94f6-450e-93b4-cbac57bd9d04.png">
https://custom-elements-manifest.netlify.app/?source=CmNsYXNzIE15RWxlbWVudCBleHRlbmRzIEhUTUxFbGVtZW50IHt9CgpjdXN0b21FbGVtZW50cy5kZWZpbmUoJ215LWVsZW1lbnQnLCBNeUVsZW1lbnQpOwo%3D&library=null

**Comment by natemoo-re** [comment]
I checked and we can definitely make this happen! It's a relatively small internal code change.

One caveat: the file that calls `customElements.define` needs to export the `MyElement` class so Astro can determine the file to load. If we do make this change, the following will work.

```js
// this must be exported!
export class MyElement extends HTMLElement {}

customElements.define("my-element", MyCustomElement)
```

```astro
---
import "../components/MyElement.js"
---

<my-element client:idle></my-element>
```

**Reply by thepassle** [reply]
Cool! Thatd already be a really nice improvement. Out of curiosity, why does the class need to be exported? Is the customElements.define call not enough to determine the file as containing a component?

**Reply by matthewp** [reply]
No, we need to know the exact file to know what to load in the client. We do that by comparing your import's exports to what component is used. If you don't export anything then that import statement can't be matched to the usage and therefore we don't know what file to load in the client.
